### ISSUE TITLE
Factors affecting max number of websocket connections

### ISSUE THREAD

**Issue by leonh** [aaugustin/websockets#772]
Hi, I have a simple chat demo that uses this websockets package via Starlette, via FastAPI. 

I have been testing how many websocket connections I can support on my local mac laptop using artillery.io's free testing tool. I can get somewhere near 300 simultaneous websocket connections working ok but ramping up to somewhere around 400-500 connection things start to seriously fall apart... both artillery and my python script hang with no indication of what the problem is. 

Is there any way I can improve the situation? What factors limit the number of connections that can be supported? How could I support more connections on a single machine? Can anyone offer any advice?

My code is here https://github.com/leonh/redis-streams-fastapi-chat

Thanks for your work on this package!

**Comment by nathanvmag** [comment]
I have same insue and dont know what do too

**Comment by aaugustin** [comment]
I already opened 10k connections to websockets on my laptop — that is 10k clients and 10k servers, so 20k sockets overall.

This requires bumping operating system limits.

Here's what I did on macOS:
`sudo sysctl -w kern.maxfiles=40960 kern.maxfilesperproc=20480`
Then in the shell before starting a script: `ulimit -n 10240`

If you can open large numbers of TCP connection with `asyncio` but you can't with `websockets`, then that would signal an issue with websockets. It's quite unlikely.

Marking as a documentation issue.

**Comment by aaugustin** [comment]
That said, it doesn't look like Starlette relies on websockets, so you may have filed this issue in the wrong issue tracker.

**Comment by leonh** [comment]
Oops! Indeed it is the wrong tracker..... but thanks for your advice anyway :-)  After looking through the docs etc. likely I will be switching to your websocket library.

**Comment by aaugustin** [comment]
You're welcome! It's still worth documenting, because other might want to make the same test with `websockets` and find the advice helpful.

**Comment by jonathan-marcus** [comment]
@aaugustin I ran a FastAPI server (based on Starlette, based on Uvicorn) in my shell, put a `breakpoint()` inside a websocket endpoint, and dug down through the many layers of abstraction to find a `websockets` socket at the bottom. 

Some evidence for this, though not a reproduction:
* Start a blank python install (I use `docker run --rm -it python:3.8`).
* `python -c 'import websockets'` fails.
* Run `pip install starlette`. Importing websockets still fails.
* Run `pip install fastapi`. Importing websockets still fails.
* Run `pip install uvicorn`. This installs `websockets` and now the import works.

I don't really understand Starlette deeply, but it appears that it is modular, and running it inside Uvicorn can make it use `websockets.`

**Comment by aaugustin** [comment]
I'm not quite sure where to document this. If we add benchmarks, that will be a good place to explain how to run benchmarks. So closing as a duplicate of #592.
